File: frontend/src/test/subscription-profile.test.ts
import { describe, expect, it } from 'vitest';

import type { StreamSettings } from '@/schemas/api/inbound';
import {
  createSubscriptionProfileDraft,
  expandSubscriptionProfileEndpoints,
} from '@/lib/xray/subscription-profile';

function baseStream(): StreamSettings {
  return {
    network: 'tcp',
    tcpSettings: {
      acceptProxyProtocol: false,
      header: { type: 'none' },
    },
    security: 'none',
  };
}

describe('subscription profile expansion', () => {
  it('creates a safe editable profile draft', () => {
    expect(createSubscriptionProfileDraft('edge.example.com', 8443)).toEqual({
      enabled: true,
      remark: '',
      dest: 'edge.example.com',
      port: 8443,
      network: 'same',
      security: 'same',
      forceTls: 'same',
      excludeFromSubTypes: [],
      mihomoX25519: false,
      shuffleHost: false,
    });
    expect(createSubscriptionProfileDraft('', 0).port).toBe(443);
  });

  it('keeps the legacy single default configuration when profiles are absent', () => {
    const endpoints = expandSubscriptionProfileEndpoints(baseStream(), 'node.example.com', 27543);

    expect(endpoints).toHaveLength(1);
    expect(endpoints[0]).toMatchObject({
      address: 'node.example.com',
      port: 27543,
      remark: '',
      profile: null,
    });
    expect(endpoints[0].streamSettings.externalProxy).toBeUndefined();
  });

  it('filters disabled profiles and applies independent WS/TLS settings', () => {
    const stream: StreamSettings = {
      ...baseStream(),
      externalProxy: [
        {
          enabled: false,
          remark: 'disabled',
          dest: 'disabled.example.com',
          port: 443,
          network: 'same',
          security: 'same',
          forceTls: 'same',
        },
        {
          enabled: true,
          remark: 'ws-tls',
          dest: 'cdn.example.com',
          port: 8443,
          network: 'ws',
          security: 'tls',
          forceTls: 'same',
          wsSettings: {
            acceptProxyProtocol: false,
            path: '/secx',
            host: 'origin.example.com',
            headers: {},
            heartbeatPeriod: 0,
          },
          tlsSettings: {
            serverName: 'sni.example.com',
            alpn: ['h2'],
            settings: {
              fingerprint: 'chrome',
              echConfigList: '',
              pinnedPeerCertSha256: [],
      verifyPeerCertByName: '',
              allowInsecure: true,
            },
          },
        },
      ],
    };

    const endpoints = expandSubscriptionProfileEndpoints(stream, 'node.example.com', 27543);

    expect(endpoints).toHaveLength(1);
    expect(endpoints[0].address).toBe('cdn.example.com');
    expect(endpoints[0].port).toBe(8443);
    expect(endpoints[0].streamSettings.network).toBe('ws');
    expect(endpoints[0].streamSettings.security).toBe('tls');
    if (endpoints[0].streamSettings.network !== 'ws') throw new Error('expected ws');
    expect(endpoints[0].streamSettings.wsSettings.path).toBe('/secx');
    expect('tcpSettings' in endpoints[0].streamSettings).toBe(false);
    if (endpoints[0].streamSettings.security !== 'tls') throw new Error('expected tls');
    expect(endpoints[0].streamSettings.tlsSettings.serverName).toBe('sni.example.com');
  });

  it('preserves a profile-specific Mux override for JSON subscription generation', () => {
    const stream: StreamSettings = {
      ...baseStream(),
      externalProxy: [
        {
          enabled: true,
          remark: 'mux',
          dest: 'mux.example.com',
          port: 443,
          network: 'same',
          security: 'same',
          forceTls: 'same',
          mux: {
            enabled: true,
            concurrency: 4,
            xudpConcurrency: 8,
            xudpProxyUDP443: 'allow',
          },
        },
      ],
    };

    const [endpoint] = expandSubscriptionProfileEndpoints(stream, 'node.example.com', 27543);
    expect(endpoint.profile?.mux).toEqual({
      enabled: true,
      concurrency: 4,
      xudpConcurrency: 8,
      xudpProxyUDP443: 'allow',
    });
  });

  it('expands two inbounds with three profiles into exactly six endpoints', () => {
    const withProfiles = (prefix: string): StreamSettings => ({
      ...baseStream(),
      externalProxy: [1, 2, 3].map((index) => ({
        enabled: true,
        remark: `${prefix}${index}`,
        dest: `${prefix}${index}.example.com`,
        port: 443,
        network: 'same' as const,
        security: 'same' as const,
        forceTls: 'same' as const,
      })),
    });

    const total = [withProfiles('a'), withProfiles('b')]
      .flatMap((stream) => expandSubscriptionProfileEndpoints(stream, 'node.example.com', 27543));
    expect(total).toHaveLength(6);
  });

  it('returns no configurations when a configured profile list is fully disabled', () => {
    const stream: StreamSettings = {
      ...baseStream(),
      externalProxy: [
        {
          enabled: false,
          remark: 'one',
          dest: 'one.example.com',
          port: 443,
          network: 'same',
          security: 'same',
          forceTls: 'same',
        },
      ],
    };

    expect(expandSubscriptionProfileEndpoints(stream, 'node.example.com', 27543)).toEqual([]);
  });

  it('keeps legacy forceTls/SNI fields working', () => {
    const stream: StreamSettings = {
      ...baseStream(),
      externalProxy: [
        {
          enabled: true,
          remark: 'legacy',
          dest: 'legacy.example.com',
          port: 443,
          network: 'same',
          security: 'same',
          forceTls: 'tls',
          sni: 'sni.example.com',
          fingerprint: 'firefox',
          alpn: ['h2'],
        },
      ],
    };

    const [endpoint] = expandSubscriptionProfileEndpoints(stream, 'node.example.com', 27543);
    expect(endpoint.streamSettings.security).toBe('tls');
    if (endpoint.streamSettings.security !== 'tls') throw new Error('expected tls');
    expect(endpoint.streamSettings.tlsSettings.serverName).toBe('sni.example.com');
    expect(endpoint.streamSettings.tlsSettings.settings.fingerprint).toBe('firefox');
  });
});

File: internal/sub/subscription_profile_test.go
package sub

import "testing"

func TestExpandSubscriptionEndpoints_DefaultWhenProfilesAbsent(t *testing.T) {
	base := map[string]any{
		"network":  "tcp",
		"security": "none",
		"tcpSettings": map[string]any{
			"header": map[string]any{"type": "none"},
		},
	}

	endpoints := expandSubscriptionEndpoints(base, "node.example.com", 443)
	if len(endpoints) != 1 {
		t.Fatalf("len(endpoints) = %d, want 1", len(endpoints))
	}
	if endpoints[0].Address != "node.example.com" || endpoints[0].Port != 443 {
		t.Fatalf("unexpected default endpoint: %#v", endpoints[0])
	}
	if _, leaked := endpoints[0].Stream["externalProxy"]; leaked {
		t.Fatal("externalProxy must not leak into the effective client stream")
	}
}

func TestExpandSubscriptionEndpoints_BlankAddressInheritsResolvedDefault(t *testing.T) {
	base := map[string]any{
		"network":  "tcp",
		"security": "none",
		"tcpSettings": map[string]any{
			"header": map[string]any{"type": "none"},
		},
		"externalProxy": []any{
			map[string]any{
				"enabled": true,
				"remark":  "empty",
				"dest":    "",
				"port":    float64(443),
			},
			map[string]any{
				"enabled": true,
				"remark":  "whitespace",
				"dest":    "   ",
				"port":    float64(8443),
			},
			map[string]any{
				"enabled": true,
				"remark":  "override",
				"dest":    "cdn.example.com",
				"port":    float64(9443),
			},
		},
	}

	endpoints := expandSubscriptionEndpoints(base, "resolved-node.example.com", 27543)
	if len(endpoints) != 3 {
		t.Fatalf("len(endpoints) = %d, want 3", len(endpoints))
	}

	wantAddresses := []string{
		"resolved-node.example.com",
		"resolved-node.example.com",
		"cdn.example.com",
	}
	for i, want := range wantAddresses {
		if endpoints[i].Address != want {
			t.Errorf("endpoints[%d].Address = %q, want %q", i, endpoints[i].Address, want)
		}
	}
}

func TestResolveExternalProxyDefaults_InheritsWithoutMutatingSource(t *testing.T) {
	original := map[string]any{
		"dest":     "   ",
		"port":     float64(0),
		"remark":   "inherit",
		"forceTls": "same",
	}
	override := map[string]any{
		"dest":     "  cdn.example.com  ",
		"port":     float64(8443),
		"remark":   "override",
		"forceTls": "same",
	}

	resolved := resolveExternalProxyDefaults(
		[]any{original, override},
		"node7.example.com",
		443,
	)
	if len(resolved) != 2 {
		t.Fatalf("len(resolved) = %d, want 2", len(resolved))
	}

	inherited := resolved[0].(map[string]any)
	explicit := resolved[1].(map[string]any)

	if inherited["dest"] != "node7.example.com" || inherited["port"] != float64(443) {
		t.Fatalf("inherited endpoint = %#v", inherited)
	}
	if explicit["dest"] != "cdn.example.com" || explicit["port"] != float64(8443) {
		t.Fatalf("explicit endpoint = %#v", explicit)
	}

	if original["dest"] != "   " || original["port"] != float64(0) {
		t.Fatalf("source profile was mutated: %#v", original)
	}
	inherited["remark"] = "changed"
	if original["remark"] != "inherit" {
		t.Fatalf("resolved profile aliases source map: %#v", original)
	}
}

func TestExpandSubscriptionEndpoints_FiltersDisabledAndOverridesStream(t *testing.T) {
	base := map[string]any{
		"network":  "tcp",
		"security": "none",
		"tcpSettings": map[string]any{
			"header": map[string]any{"type": "none"},
		},
		"externalProxy": []any{
			map[string]any{
				"enabled": false,
				"remark":  "disabled",
				"dest":    "disabled.example.com",
				"port":    float64(443),
			},
			map[string]any{
				"enabled":  true,
				"remark":   "ws-tls",
				"dest":     "cdn.example.com",
				"port":     float64(8443),
				"network":  "ws",
				"security": "tls",
				"wsSettings": map[string]any{
					"path":            "/secx",
					"host":            "origin.example.com",
					"headers":         map[string]any{},
					"heartbeatPeriod": float64(0),
				},
				"tlsSettings": map[string]any{
					"serverName": "sni.example.com",
					"alpn":       []any{"h2"},
					"settings": map[string]any{
						"fingerprint":   "chrome",
						"allowInsecure": true,
					},
				},
			},
		},
	}

	endpoints := expandSubscriptionEndpoints(base, "node.example.com", 27543)
	if len(endpoints) != 1 {
		t.Fatalf("len(endpoints) = %d, want 1 active profile", len(endpoints))
	}
	endpoint := endpoints[0]
	if endpoint.Address != "cdn.example.com" || endpoint.Port != 8443 || endpoint.Remark != "ws-tls" {
		t.Fatalf("unexpected endpoint: %#v", endpoint)
	}
	if endpoint.Stream["network"] != "ws" || endpoint.Stream["security"] != "tls" {
		t.Fatalf("effective stream did not apply network/security: %#v", endpoint.Stream)
	}
	if _, stillPresent := endpoint.Stream["tcpSettings"]; stillPresent {
		t.Fatal("old transport settings must be removed when network changes")
	}
	ws, _ := endpoint.Stream["wsSettings"].(map[string]any)
	if ws["path"] != "/secx" || ws["host"] != "origin.example.com" {
		t.Fatalf("unexpected ws settings: %#v", ws)
	}
	tlsSettings, _ := endpoint.Stream["tlsSettings"].(map[string]any)
	if tlsSettings["serverName"] != "sni.example.com" {
		t.Fatalf("unexpected TLS settings: %#v", tlsSettings)
	}
}

func TestExpandSubscriptionEndpoints_AllDisabledMeansNoConfiguration(t *testing.T) {
	base := map[string]any{
		"network":  "tcp",
		"security": "none",
		"tcpSettings": map[string]any{
			"header": map[string]any{"type": "none"},
		},
		"externalProxy": []any{
			map[string]any{"enabled": false, "dest": "one.example.com", "port": float64(443)},
			map[string]any{"enabled": false, "dest": "two.example.com", "port": float64(8443)},
		},
	}

	if endpoints := expandSubscriptionEndpoints(base, "node.example.com", 443); len(endpoints) != 0 {
		t.Fatalf("len(endpoints) = %d, want 0", len(endpoints))
	}
}

func TestExpandSubscriptionEndpoints_LegacyTLSFieldsRemainCompatible(t *testing.T) {
	base := map[string]any{
		"network":  "tcp",
		"security": "none",
		"tcpSettings": map[string]any{
			"header": map[string]any{"type": "none"},
		},
		"externalProxy": []any{
			map[string]any{
				"forceTls":    "tls",
				"dest":        "legacy.example.com",
				"port":        float64(443),
				"remark":      "legacy",
				"sni":         "sni.example.com",
				"fingerprint": "firefox",
				"alpn":        []any{"h2", "http/1.1"},
			},
		},
	}

	endpoints := expandSubscriptionEndpoints(base, "node.example.com", 27543)
	if len(endpoints) != 1 {
		t.Fatalf("len(endpoints) = %d, want 1", len(endpoints))
	}
	stream := endpoints[0].Stream
	if stream["security"] != "tls" {
		t.Fatalf("security = %v, want tls", stream["security"])
	}
	tlsSettings, _ := stream["tlsSettings"].(map[string]any)
	if tlsSettings["serverName"] != "sni.example.com" {
		t.Fatalf("serverName = %v, want sni.example.com", tlsSettings["serverName"])
	}
	clientSettings, _ := tlsSettings["settings"].(map[string]any)
	if clientSettings["fingerprint"] != "firefox" {
		t.Fatalf("fingerprint = %v, want firefox", clientSettings["fingerprint"])
	}
}

func TestExpandSubscriptionEndpoints_CarriesPerProfileMuxOverride(t *testing.T) {
	base := map[string]any{
		"network":  "tcp",
		"security": "none",
		"tcpSettings": map[string]any{
			"header": map[string]any{"type": "none"},
		},
		"externalProxy": []any{
			map[string]any{
				"enabled": true,
				"dest":    "mux.example.com",
				"port":    float64(443),
				"mux": map[string]any{
					"enabled":         true,
					"concurrency":     float64(4),
					"xudpConcurrency": float64(8),
					"xudpProxyUDP443": "allow",
				},
			},
		},
	}

	endpoints := expandSubscriptionEndpoints(base, "node.example.com", 27543)
	if len(endpoints) != 1 || !endpoints[0].HasMuxOverride {
		t.Fatalf("unexpected endpoints/mux override: %#v", endpoints)
	}
	mux, ok := endpoints[0].MuxOverride.(map[string]any)
	if !ok || mux["enabled"] != true || mux["concurrency"] != float64(4) {
		t.Fatalf("unexpected mux override: %#v", endpoints[0].MuxOverride)
	}
}

func TestExpandSubscriptionEndpoints_TwoInboundsWithThreeProfilesProduceSixConfigurations(t *testing.T) {
	makeStream := func(prefix string) map[string]any {
		profiles := make([]any, 0, 3)
		for i := 1; i <= 3; i++ {
			profiles = append(profiles, map[string]any{
				"enabled": true,
				"remark":  prefix + string(rune('0'+i)),
				"dest":    prefix + string(rune('0'+i)) + ".example.com",
				"port":    float64(443),
			})
		}
		return map[string]any{
			"network":       "tcp",
			"security":      "none",
			"tcpSettings":   map[string]any{"header": map[string]any{"type": "none"}},
			"externalProxy": profiles,
		}
	}

	total := 0
	for _, stream := range []map[string]any{makeStream("a"), makeStream("b")} {
		total += len(expandSubscriptionEndpoints(stream, "node.example.com", 27543))
	}
	if total != 6 {
		t.Fatalf("total configurations = %d, want exactly 6", total)
	}
}
